src/fldmod_by_const.jl

```diff
@@ -0,0 +1,147 @@
+function splitint(x::T) where {T<:Integer}
+    NT = narrow(T)
+    (x >> nbits(NT)) % NT, x % NT
+end
+
+narrow(::Type{Int128}) = Int64
+narrow(::Type{Int64}) = Int32
+narrow(::Type{Int32}) = Int16
+narrow(::Type{Int16}) = Int8
+narrow(::Type{UInt128}) = UInt64
+narrow(::Type{UInt64}) = UInt32
+narrow(::Type{UInt32}) = UInt16
+narrow(::Type{UInt16}) = UInt8
+
+"""    nbits(x) = sizeof(x)*8 """
+nbits(x) = sizeof(x)*8
+
+# Implemenation based on umul32hi, from https://stackoverflow.com/a/22847373/751061
+# /* compute the upper 32 bits of the product of two unsigned 32-bit integers */
+@inline function unsigned_splitmul_upper(a::T, b::T) where T<:Unsigned
+    # /* split operands into halves */
+    ah,al = splitint(a)
+    bh,bl = splitint(b)
+    halfT = typeof(ah)
+    halfbits = nbits(al)
+    # /* compute partial products */
+    p0 = widemul(al, bl);
+    p1 = widemul(al, bh);
+    p2 = widemul(ah, bl);
+    p3 = widemul(ah, bh);
+    # /* sum partial products */
+    cy = ((p0 >> halfbits) + (p1%halfT) + (p2%halfT)) >> halfbits;
+    return p3 + (p2 >> halfbits) + (p1 >> halfbits) + cy;
+end
+@inline function splitmul_upper(a::T, b::T) where T<:Unsigned
+    return unsigned_splitmul_upper(a,b)
+end
+@inline function splitmul_upper(a::T, b::T) where T<:Signed
+    return signed(unsigned_splitmul_upper(unsigned(a), unsigned(b))) - ((a < 0) ? b : 0) - ((b < 0) ? a : 0);
+end
+@inline splitmul_upper(a,b) = splitmul_upper(promote(a,b)...)
+
+
+"""
+    div_by_const(x, Val(C))
+
+The idea behind `div_by_const(x, Val(C))` is that we can avoid division entirely if
+instead of dividing by `C`, we multiply by `(2^N/C)`, then "divide by" `2^N`.
+
+And if we choose `N` to be `nbits(T)`, we can avoid the "divide by `2^N`" by just taking the
+_upper half_ of the result of the multiplication. And so, to do the multiplication, we use
+a `splitwidemul` which avoids widening by doing splitting x into upper and lower bits, doing
+several smaller multiplies, and then returning the result as two Ts, for the up and lo.
+
+REQUIRES:
+  - `C` _must be_ greater than `0`
+"""
+function div_by_const(x::T, ::Val{C}) where {T, C}
+# These checks will be compiled away during specialization.
+    if C == 1
+        return x
+    elseif (C <= 0)
+        throw(DomainError("C must be > 0"))
+    end
+    inv_coeff, toshift = calculate_inv_coeff(T, C)
+    up = splitmul_upper(x, inv_coeff)
+    out = up    # By keeping only the upper half, we're essentially dividing by 2^nbits(T)
+    # This condition will be compiled away during specialization.
+    if T <: Signed
+        # Because our magic number has a leading one, the result is negative if it's Signed.
+        # We add x to give us the positive equivalent.
+        out += x
+        signshift = (nbits(x)-1)
+        signed = T(unsigned(out) >> signshift)  # "unsigned" bitshift (to read top bit)
+    end
+    out = out >> toshift
+    if T <: Signed
+        out =  out + signed     # ; leaq    (%rdx,%rax), %rax
```

Comment:
Is the compiler smart enough to see that 'signed' is definitely assigned here?

Reply:
Yes. It's hard to tell, but since `if T <: Signed` is "statically" determinable, this is almost like a `#ifdef` type check -- these blocks are either removed or not removed from the final code that's generated:

```julia
julia> @code_native FixedPointDecimals.div_by_const(Int64(232), Val(100))
         ...
    sarq    $6, %rax
    leaq    (%rax,%rcx), %rax
    retq

julia> @code_native FixedPointDecimals.div_by_const(UInt64(232), Val(100))
        ...
    shrq    $6, %rax
    retq
```

Reply:
You can also see in the `@code_typed` that the compiler gives `x` a `MaybeUndef` value:
```julia
114 16 ─ %69 = φ (#15 => %65, #9 => #undef)::MaybeUndef(Int64)                                         │
```
but `#9` is a compiler const `true`, so it'll know that it is always reachable:
```julia
107 9 ── %51 = ($(Expr(:static_parameter, 1)) <: Main.FixedPointDecimals.Signed)::Const(true, false)   │
    └───       goto #16 if not %51                                                                     │
```